{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Fm_wq8sBkiS3"
      },
      "source": [
        "# ColabAlign\n",
        "\n",
        "### Fast pairwise protein secondary structure comparisons using multiprocessing\n",
        "\n",
        "[![DOI](https://zenodo.org/badge/788453062.svg)](https://doi.org/10.5281/zenodo.14169501)\n",
        "\n",
        "Create a phylogenetic tree that compares the secondary structure of proteins, rather than nucleotide or amino acid sequence. Scoring uses the [_US-align_](https://zhanggroup.org/US-align/) algorithm by [_Zhang et al. (2022)_](https://doi.org/10.1038/s41592-022-01585-1). Based on [_mTM-align_](http://yanglab.nankai.edu.cn/mTM-align/) by [_Dong et al. (2018)_](https://doi.org/10.1093/nar/gky430). Made possible with [US-align](https://github.com/pylelab/USalign) and [Biopython](https://biopython.org).\n",
        "\n",
        "A score of **<0.17** indicates similarity indistinguishable from a random pair of structures, where as as score **≥0.50** indicates a pair with broadly the same fold ([_Xu et al., 2010_](https://doi.org/10.1093/bioinformatics/btq066))\n",
        "\n",
        "##### Usage\n",
        "\n",
        "* For the best performance, click `Runtime` -> `Change runtime type` -> `TPU v2`\n",
        "\n",
        "* If `TPU v2` is not available, `L4 GPU` or `T4 GPU` runtimes (paid options) are the next best options. These perform about 10x faster than the default `CPU` runtime. I recommend using `TPU v2`, `L4 GPU` or `T4 GPU` for datasets containing > 250 structures.\n",
        "\n",
        "* Upload .pdb or .cif format files directly to the Colab instance by clicking the folder icon on the left, then dragging and dropping your structures.\n",
        "\n",
        "* US-align only considers the first chain in each .pdb or .cif file, so please ensure this is the chain you wish to include in the pairwise alignment.\n",
        "\n",
        "* Phylogenetic trees generated by ColabAlign should be viewed/analysed as un-rooted trees."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "e8POztA5W7Qa",
        "outputId": "faaf30f5-02b6-44f0-ab23-f43654cf4837",
        "cellView": "form"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.11/dist-packages (3.10.0)\n",
            "Collecting treecluster\n",
            "  Downloading treecluster-1.0.4-py2.py3-none-any.whl.metadata (590 bytes)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (2.2.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (1.26.4)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (1.3.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (4.56.0)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (1.4.8)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (24.2)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (11.1.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (3.2.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (2.9.0.post0)\n",
            "Collecting niemads (from treecluster)\n",
            "  Downloading niemads-1.0.16-py2.py3-none-any.whl.metadata (860 bytes)\n",
            "Collecting treeswift (from treecluster)\n",
            "  Downloading treeswift-1.1.45-py2.py3-none-any.whl.metadata (1.2 kB)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.7->matplotlib) (1.17.0)\n",
            "Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m29.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading treecluster-1.0.4-py2.py3-none-any.whl (18 kB)\n",
            "Downloading niemads-1.0.16-py2.py3-none-any.whl (18 kB)\n",
            "Downloading treeswift-1.1.45-py2.py3-none-any.whl (33 kB)\n",
            "Installing collected packages: treeswift, niemads, treecluster, biopython\n",
            "Successfully installed biopython-1.85 niemads-1.0.16 treecluster-1.0.4 treeswift-1.1.45\n",
            "Cloning into 'USalign_git'...\n",
            "remote: Enumerating objects: 1542, done.\u001b[K\n",
            "remote: Counting objects: 100% (407/407), done.\u001b[K\n",
            "remote: Compressing objects: 100% (55/55), done.\u001b[K\n",
            "remote: Total 1542 (delta 381), reused 369 (delta 352), pack-reused 1135 (from 2)\u001b[K\n",
            "Receiving objects: 100% (1542/1542), 798.69 KiB | 6.19 MiB/s, done.\n",
            "Resolving deltas: 100% (1103/1103), done.\n",
            "--2025-02-26 12:44:06--  https://raw.githubusercontent.com/crfield18/ColabAlign/refs/heads/main/colabalign.py\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 21351 (21K) [text/plain]\n",
            "Saving to: ‘colabalign.py’\n",
            "\n",
            "colabalign.py       100%[===================>]  20.85K  --.-KB/s    in 0.002s  \n",
            "\n",
            "2025-02-26 12:44:06 (12.5 MB/s) - ‘colabalign.py’ saved [21351/21351]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# @title Download and Compile the latest version of US-align, Biopython and pandas\n",
        "\n",
        "import os\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display\n",
        "\n",
        "threshold = 0.1 # @param {type:\"slider\", min:0.01, max:1.00, step:0.01}\n",
        "\n",
        "os.environ['threshold'] = str(threshold)\n",
        "\n",
        "!pip install biopython matplotlib treecluster pandas numpy\n",
        "\n",
        "# Download and compile an up-to-date copy of US-align\n",
        "!git clone https://github.com/pylelab/USalign.git USalign_git\n",
        "!g++ -static -O3 -ffast-math -lm -o USalign USalign_git/USalign.cpp\n",
        "\n",
        "!wget https://raw.githubusercontent.com/crfield18/ColabAlign/refs/heads/main/colabalign.py -O colabalign.py\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kFZ3zmeFhP7n"
      },
      "outputs": [],
      "source": [
        "# @title Align Structures\n",
        "\n",
        "# @markdown Make sure your .pdb and/or .cif files are uploaded before running this block.\n",
        "\n",
        "%%script bash\n",
        "\n",
        "mkdir -p results models\n",
        "\n",
        "tar -xzf *tar.gz > /dev/null 2>&1\n",
        "\n",
        "# Delete any PDB or mmCIF files downloaded from the US-align repository.\n",
        "# US-align comes with 2 example PDB files and we don't want to add these to our tree.\n",
        "rm USalign_git/*.pdb\n",
        "\n",
        "# Move all pdb and cif files to the models directory to make file handling with colabalign.py a lot easier\n",
        "find . -type f \\( -name \"*.pdb\" -o -name \"*.cif\" \\) | tee /content/results/models.txt | xargs -I {} mv {} models/ > /dev/null 2>&1\n",
        "\n",
        "# Count the number of models in models/ to show how many alignments are needed.\n",
        "# Mainly to estimate how long the job will take.\n",
        "model_count=$(find models/ -type f | wc -l)\n",
        "calc_count=$((($model_count * ($model_count - 1)) / 2))\n",
        "echo \"There are ${model_count} files in the models directory. This means ${calc_count} total calculations.\"\n",
        "\n",
        "# Run colabalign.py with all available cores for the current runtype.\n",
        "python3 colabalign.py -i /content/models -o /content/results -c $(nproc --all) -t ${threshold} -p /content/USalign\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "e1XfdKzakiS7"
      },
      "outputs": [],
      "source": [
        "# @title Zip and download results\n",
        "\n",
        "import os\n",
        "import datetime\n",
        "from google.colab import files\n",
        "\n",
        "# Name the zipped results file using the current date and time\n",
        "# to not accidentally overwrite older results files when downloading\n",
        "current_dt = datetime.datetime.now()\n",
        "zip_filename = f'colabalign_results_{current_dt.strftime(\"%Y%m%d-%H%M\")}.zip'\n",
        "\n",
        "# Using the built-in zip function rather than a python module for improved efficiency\n",
        "os.system(f'zip -r {zip_filename} results')\n",
        "files.download(zip_filename)\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "TPU",
    "colab": {
      "gpuType": "V28",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "colabalign",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.12.7"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}