{"cells":[{"cell_type":"markdown","metadata":{"id":"Fm_wq8sBkiS3"},"source":["### ColabAlign&nbsp;&nbsp;&nbsp;&nbsp;[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/crfield18/ColabAlign/blob/main/colabalign.ipynb)\n","\n","Create a phylogenetic tree that compares the secondary structure of proteins, rather than nucleotide or amino acid sequence. Scoring uses the [_TM-align_](https://zhanggroup.org/TM-align/) algorithm by [_Zhang and Skolnick (2005)_](https://doi.org/10.1093%2Fnar%2Fgki524). Based of [_mTM-align_](http://yanglab.nankai.edu.cn/mTM-align/) by [_Dong et al. (2018)_](https://doi.org/10.1093/nar/gky430). Made possible with [US-align](https://github.com/pylelab/USalign) and [Biopython](https://biopython.org).\n","\n","A score of **<0.17** indicates similarity indistinguishable from a random pair of structures, where as as score **â‰¥0.50** indicates a pair with broadly the same fold ([_Xu et al., 2010_](https://doi.org/10.1093/bioinformatics/btq066))\n","\n","\n","##### Usage\n","\n","* For the best performance, click `Runtime` -> `Change runtime type` -> `TPU v2`\n","\n","* Upload .pdb or .cif format files directly to the Colab instance by clicking the folder icon on the left, then dragging and dropping your structures\n","\n","* For the best results, upload PDB/mmCIF files that contain a single subunit and with only a single position for each atom (crystallography/cryo-EM structures from the PDB can sometimes have multiple positions for some residues where a single position could not be resolved).\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"e8POztA5W7Qa"},"outputs":[],"source":["# @title Download and Compile the latest version of TM-align, Biopython and pandas\n","\n","%%script bash\n","\n","pip install biopython matplotlib\n","# pip install pandas # pandas is already installed on Colab. Uncomment if this changes\n","\n","# Download and compile an up-to-date copy of TM-align/US-align\n","git clone https://github.com/pylelab/USalign.git\n","g++ -static -O3 -ffast-math -lm -o TMalign USalign/TMalign.cpp"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"kFZ3zmeFhP7n"},"outputs":[],"source":["# @title Set up file structure\n","\n","# @markdown Make sure your .pdb and/or .cif files are uploaded before running this block.\n","\n","%%script bash\n","\n","mkdir -p results models results/sub-lists results\n","\n","# Delete any PDB or mmCIF files downloaded from the US-align (TM-align) repository\n","# US-align comes with 2 example PDB files and we don't want to add these to our\n","rm USalign/*.pdb\n","\n","find . -type f \\( -name \"*.pdb\" -o -name \"*.cif\" \\) -exec mv {} models/ \\; > /dev/null 2>&1\n","\n","# Count the number of models in models/ to show how many alignments are needed\n","# Mainly to estimate how long\n","model_count=$(find models/ -type f | wc -l)\n","calc_count=$((($model_count * ($model_count - 1)) / 2))\n","echo \"There are ${model_count} files in the models directory. This means ${calc_count} total calculations.\"\n","\n","# Make a list of all PDB and mmCIF files in the models directory and write to model_list.txt in the results directory\n","# model_list.txt is used to split work up between threads later\n","find models/ -maxdepth 1 \\( -name \"*.pdb\" -o -name \"*.cif\" \\) -printf \"%f\\n\" > results/model_list.txt"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"R7_g8hoBxOdO"},"outputs":[],"source":["# @title Run pairwise TM-align calculations\n","\n","import os\n","import subprocess\n","from itertools import combinations\n","from concurrent.futures import ProcessPoolExecutor, as_completed\n","\n","# Read model list from file\n","with open('results/model_list.txt', 'r') as model_list_file:\n","  model_list = [line.strip() for line in model_list_file]\n","\n","# print(f'Core count:\\t{os.cpu_count()}')\n","\n","# Make sub-lists of models to compare using combinations\n","all_combos = combinations(model_list, 2)\n","\n","sub_lists = {}\n","for pair in all_combos:\n","  if pair[0] not in sub_lists:\n","    sub_lists[pair[0]] = []\n","  sub_lists[pair[0]].append(pair[1])\n","\n","# Write individual model list files for each target model\n","for target, sub_list in sub_lists.items():\n","  with open(f'results/sub-lists/{target}-model_list.txt', 'w') as sublist_file:\n","    for model in sub_list:\n","      sublist_file.writelines(f'{model}\\n')\n","\n","# Function to run TM-align for a given model\n","def run_tmalign(model):\n","  print(f'Running TM-align for {model}')\n","  # Align model against models listed in corresponding model list file\n","  cmd = f'./TMalign models/{model} -dir2 models/ results/sub-lists/{model}-model_list.txt -outfmt 2 > results/sub-lists/{model}-tm-align_out.tsv'\n","  process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n","  stdout, stderr = process.communicate()\n","  return model, stdout, stderr\n","\n","# Run multiple TM-align jobs in parallel\n","with ProcessPoolExecutor(max_workers=os.cpu_count()) as executor:\n","  # Submit all tasks to the executor\n","  futures = [executor.submit(run_tmalign, model) for model in sub_lists.keys()]\n","\n","  # Ensure all tasks are completed\n","  for future in as_completed(futures):\n","    try:\n","      model, stdout, stderr = future.result()\n","      if stderr:\n","        print(f\"Error for models {model}: {stderr.decode()}\")\n","    except Exception as e:\n","      print(f\"Error occurred: {e}\")\n","\n","# Merge individual TM-align results files into 1 file\n","tmalign_results_files = [file for file in os.listdir('results/sub-lists') if file.endswith('.tsv')]\n","\n","with open('results/tm-align_out.tsv', 'w') as merged_tmalign_results_file:\n","  # Write header line\n","  merged_tmalign_results_file.write('#PDBchain1\\tPDBchain2\\tTM1\\tTM2\\tRMSD\\tID1\\tID2\\tIDali\\tL1\\tL2\\tLali\\n')\n","\n","  # Open each results file and copy all lines except the header and footer (lines beginning with #)\n","  for file in tmalign_results_files:\n","    with open(os.path.join('results/sub-lists', file), 'r') as result_file:\n","      for line in result_file:\n","        if not line.startswith('#'):\n","          # merged_tmalign_results_file.write(line[7:]) # temp fix for index naming\n","\n","          # Remove models/ from #PDBchain1 name\n","          new_line = line.split('\\t')\n","          new_line[0] = new_line[0].split('/')[-1]\n","          merged_tmalign_results_file.write('\\t'.join(new_line))"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"_EInl51QhP7n"},"outputs":[],"source":["# @title Make phylogenetic tree from TM-align results\n","\n","# %matplotlib inline\n","\n","from math import isnan\n","from pathlib import Path\n","import pandas as pd\n","import numpy as np\n","\n","from Bio import Phylo\n","from Bio.Phylo.TreeConstruction import DistanceTreeConstructor, DistanceMatrix\n","\n","class ColabAlign():\n","  def __init__(self) -> None:\n","    self.home_path = Path.cwd()\n","    self.models_path = self.home_path.joinpath('models')\n","    self.results_path = self.home_path.joinpath('results')\n","    self.colabalign = None\n","\n","    with open(self.results_path.joinpath('model_list.txt'), 'r') as model_list_file:\n","      self.model_list = [line.strip('\\n') for line in model_list_file]\n","\n","  def tmalign_output_to_matrix(self):\n","    # inputs: \n","    # outputs:  \n","\n","    tmalign_output = pd.read_csv(self.results_path.joinpath('tm-align_out.tsv'), sep='\\t')\n","\n","    # Initialize a matrix with NaN values to be populated with TM-align values\n","    # Probably not the most memory/time efficient method for this but I found it much easier\n","    # to understand how the data is handled\n","    matrix_data = np.full((len(self.model_list), len(self.model_list)), np.nan)\n","\n","    # Create a dictionary to map model names to indices\n","    model_index = {model: idx for idx, model in enumerate(self.model_list)}\n","\n","    # Populate the matrix with the highest TM-align score for each pair\n","    for index, row in tmalign_output.iterrows():\n","      if row['#PDBchain1'].startswith('#'):\n","        continue\n","      else:\n","        max_tm = max(row['TM1'], row['TM2'])\n","        i, j = model_index[row['#PDBchain1']], model_index[row['PDBchain2']]\n","        matrix_data[i, j] = max_tm\n","        matrix_data[j, i] = max_tm\n","\n","    # Fill the diagonal with scores of 1\n","    # We don't need to align a model to itself because this will always return a TM-score of 1\n","    # Calculated TM-align scores are given to 4 decimal places\n","    np.fill_diagonal(matrix_data, 1.0000)\n","\n","    self.colabalign = pd.DataFrame(matrix_data, index=self.model_list, columns=self.model_list)\n","    self.colabalign.to_csv(self.results_path.joinpath('tm-align_score_matrix.csv'), index=True)\n","\n","    return self.colabalign\n","\n","  def make_tree_from_matrix(self):\n","    # Invert TM-align scores to make them suitable for distances on a phylogenetic tree\n","    # More similar pairs of models (i.e., higher TM-scores) have shorter distances to each other\n","    # Distance values are all rounded to 4 decimal places since all TM-align scores are also to 4 d.p.\n","    # Clipping is also applied to ensure no values below 0 or above 1 are present in the distance matrix\n","    if not self.colabalign.empty:\n","      distances_df = 1.0000 - self.colabalign\n","      distances_df = distances_df.round(4)\n","      distances_df = distances_df.clip(lower=0.0000,upper=1.0000)\n","\n","    # Convert the scores matrix to a lower triangle matrix\n","    # The lower and upper triangles of the matrix are identical and Bio.Phylo.TreeConstruction\n","    # requires a lower triangle matrix rather than the full matrix\n","    lower_tri_df = distances_df.where(np.tril(np.ones(distances_df.shape)).astype(bool))\n","    lower_tri_lists = [[value for value in row if not isnan(value)] for row in lower_tri_df.values.tolist()]\n","\n","    # Generate phylogenetic tree using the UPGMA clustering method\n","    # We can safely ignore the Molecular Clock hypothesis because we are not\n","    # deriving evolutionary relationships between proteins\n","    tm_matrix = DistanceMatrix(names=self.model_list, matrix=lower_tri_lists)\n","    constructor = DistanceTreeConstructor()\n","    tree = constructor.upgma(tm_matrix)\n","\n","    print('\\n')\n","    Phylo.draw_ascii(tree)\n","    Phylo.write(tree, self.results_path.joinpath('colabalign.tree'), 'newick')\n","\n","\n","if __name__ == '__main__':\n","  colab_instance = ColabAlign()\n","  colab_instance.tmalign_output_to_matrix()\n","  colab_instance.make_tree_from_matrix()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"e1XfdKzakiS7"},"outputs":[],"source":["# @title Zip and download results\n","\n","import datetime\n","from google.colab import files\n","import os\n","\n","# Name the zipped results file using the current date and time\n","# to not accidentally overwrite older results files when downloading\n","current_dt = datetime.datetime.now()\n","zip_filename = f'colabalign_results_{current_dt.strftime(\"%Y%m%d-%H%M\")}.zip'\n","\n","# Using the built-in zip function rather than a python module for efficiency\n","os.system(f'zip -r {zip_filename} models results ')\n","files.download(zip_filename)"]}],"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/crfield18/ColabAlign/blob/main/colabalign.ipynb","timestamp":1723733455516},{"file_id":"https://github.com/crfield18/TMmatrix/blob/main/tmmatrix.ipynb","timestamp":1722006336643}]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.12.2"}},"nbformat":4,"nbformat_minor":0}
