{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Fm_wq8sBkiS3"
      },
      "source": [
        "# **ColabAlign**\n",
        "\n",
        "## Fast pairwise protein secondary structure comparisons using multiprocessing\n",
        "\n",
        " [![Code DOI](https://img.shields.io/badge/DOI-10.5281/zenodo.14169501-blue)](https://doi.org/10.5281/zenodo.14169501) [![Paper DOI](https://img.shields.io/badge/DOI-10.1101/2025.10.06.677802-green)](https://doi.org/10.1101/2025.10.06.677802)\n",
        "\n",
        "---\n",
        "\n",
        "Create a phylogenetic tree that compares the secondary structure of proteins, rather than nucleotide or amino acid sequence. Scoring uses the [US-align](https://zhanggroup.org/US-align/) algorithm by [Zhang _et al._, (2022)](https://doi.org/10.1038/s41592-022-01585-1).\n",
        "\n",
        "A score of **<0.17** indicates similarity indistinguishable from a random pair of structures, where as as score **â‰¥0.50** indicates a pair with broadly the same fold ([Xu _et al._, 2010](https://doi.org/10.1093/bioinformatics/btq066))\n",
        "\n",
        "---\n",
        "\n",
        "## **Usage**\n",
        "\n",
        "1) **Click `Runtime` -> `Change runtime type` -> `v5e-1 TPU`**\n",
        "\n",
        "2) **Either upload `.pdb` and/or `.cif` format files directly or compressed into a `.tar.gz` file to the Colab instance by clicking the folder icon on the left, then dragging and dropping your structures.**\n",
        "\n",
        "3) **Adjust the clustering threshold (horizontal bar) as needed.**\n",
        "\n",
        "4) **Click `Run all`**\n",
        "\n",
        "## N.B.\n",
        "\n",
        "* Other runtimes can be used if you are paying for compute units. Ensure you select a runtime with more than 2 CPU cores, particularly if you plan to analyse a large dataset.\n",
        "\n",
        "* US-align only considers the first chain in each .pdb or .cif file, so please ensure this is the chain you wish to include in the pairwise alignment.\n",
        "\n",
        "* Structural dendrograms generated by ColabAlign should be viewed/analysed as unrooted trees.\n",
        "\n",
        "* Clustering threshold is equivalent to 1 - the minimum TM-score for a pair of structures -- _e.g._ a threshold of 0.25 will group together structures with TM-scores >= 0.75\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "v_I2HlVXBoum",
        "outputId": "25799b8f-b161-405c-a2ab-3179c3c54a16"
      },
      "outputs": [],
      "source": [
        "# @title Set up ColabAlign conda environment and prepare file structure\n",
        "\n",
        "%%bash\n",
        "\n",
        "rm -r sample_data\n",
        "\n",
        "git clone https://github.com/crfield18/ColabAlign.git ColabAlign_git\n",
        "mv ColabAlign_git/*.py .\n",
        "mv ColabAlign_git/*.yml .\n",
        "rm -rd ColabAlign_git\n",
        "\n",
        "pip install konda\n",
        "konda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main\n",
        "konda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r\n",
        "konda env update -n base -f colabalign.yml --quiet\n",
        "\n",
        "# The version on MView in the bioconda channel doesn't currently like working on Colab, so\n",
        "# we need to manually install it instead\n",
        "git clone https://github.com/crfield18/MView.git MView_git\n",
        "cd MView_git\n",
        "perl install.pl\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OxAFOwQ-Boum",
        "outputId": "58ae72d2-7524-47c9-8bc5-90aaf52ff521"
      },
      "outputs": [],
      "source": [
        "# @title Align Structures\n",
        "# Run colabalign.py with all available cores\n",
        "\n",
        "threshold = 0.25 # @param {type:\"slider\", min:0.01, max:1.00, step:0.01}\n",
        "\n",
        "!mkdir -p results models\n",
        "!find . -name \"*.tar.gz\" -exec tar -xzf {} \\; > /dev/null 2>&1\n",
        "!find . -type f \\( -name \"*.pdb\" -o -name \"*.cif\" \\) -not -path \"./models/*\" | xargs -I {} mv {} models/ 2>/dev/null || true\n",
        "!find ./models/ -name \"._*\" -delete 2>/dev/null\n",
        "\n",
        "!echo \"Starting alignment with $(nproc --all) cores...\"\n",
        "!python3 colabalign.py -i /content/models -o /content/results -c $(nproc --all) -t $threshold -b /content/BeEM\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "e1XfdKzakiS7"
      },
      "outputs": [],
      "source": [
        "# @title Zip and download results\n",
        "\n",
        "import os\n",
        "import datetime\n",
        "from google.colab import files\n",
        "\n",
        "# Name the zipped results file using the current date and time\n",
        "# to not accidentally overwrite older results files when downloading\n",
        "current_dt = datetime.datetime.now()\n",
        "zip_filename = f'colabalign_results_{current_dt.strftime(\"%Y%m%d-%H%M\")}.zip'\n",
        "\n",
        "# Using the built-in zip function rather than a python module for improved efficiency\n",
        "os.system(f'zip -r {zip_filename} results')\n",
        "files.download(zip_filename)\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "TPU",
    "colab": {
      "gpuType": "V5E1",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.12.7"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
