{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Fm_wq8sBkiS3"
      },
      "source": [
        "# ColabAlign\n",
        "\n",
        "### Fast pairwise protein secondary structure comparisons using multiprocessing\n",
        "\n",
        "[![DOI](https://zenodo.org/badge/788453062.svg)](https://doi.org/10.5281/zenodo.14169501)\n",
        "\n",
        "Create a phylogenetic tree that compares the secondary structure of proteins, rather than nucleotide or amino acid sequence. Scoring uses the [_US-align_](https://zhanggroup.org/US-align/) algorithm by [_Zhang et al. (2022)_](https://doi.org/10.1038/s41592-022-01585-1). Based on [_mTM-align_](http://yanglab.nankai.edu.cn/mTM-align/) by [_Dong et al. (2018)_](https://doi.org/10.1093/nar/gky430). Made possible with [US-align](https://github.com/pylelab/USalign) and [Biopython](https://biopython.org).\n",
        "\n",
        "A score of **<0.17** indicates similarity indistinguishable from a random pair of structures, where as as score **â‰¥0.50** indicates a pair with broadly the same fold ([_Xu et al., 2010_](https://doi.org/10.1093/bioinformatics/btq066))\n",
        "\n",
        "##### Usage\n",
        "\n",
        "* For the best performance, click `Runtime` -> `Change runtime type` -> `TPU v2`\n",
        "\n",
        "* If `TPU v2` is not available, `L4 GPU` or `T4 GPU` runtimes (paid options) are the next best options. These perform about 10x faster than the default `CPU` runtime. I recommend using `TPU v2`, `L4 GPU` or `T4 GPU` for datasets containing > 250 structures.\n",
        "\n",
        "* Upload .pdb or .cif format files directly to the Colab instance by clicking the folder icon on the left, then dragging and dropping your structures.\n",
        "\n",
        "* US-align only considers the first chain in each .pdb or .cif file, so please ensure this is the chain you wish to include in the pairwise alignment.\n",
        "\n",
        "* Phylogenetic trees generated by ColabAlign should be viewed/analysed as un-rooted trees."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "e8POztA5W7Qa",
        "outputId": "faaf30f5-02b6-44f0-ab23-f43654cf4837"
      },
      "outputs": [],
      "source": [
        "# @title Download and Compile the latest version of US-align, Biopython and pandas\n",
        "\n",
        "import os\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display\n",
        "\n",
        "threshold = 0.1 # @param {type:\"slider\", min:0.01, max:1.00, step:0.01}\n",
        "\n",
        "os.environ['threshold'] = str(threshold)\n",
        "\n",
        "!pip install biopython matplotlib treecluster pandas numpy\n",
        "\n",
        "# Download and compile an up-to-date copy of US-align\n",
        "!git clone https://github.com/pylelab/USalign.git USalign_git\n",
        "!g++ -static -O3 -ffast-math -lm -o USalign USalign_git/USalign.cpp\n",
        "\n",
        "!wget https://raw.githubusercontent.com/crfield18/ColabAlign/refs/heads/main/colabalign.py -O colabalign.py\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kFZ3zmeFhP7n"
      },
      "outputs": [],
      "source": [
        "# @title Align Structures\n",
        "\n",
        "# @markdown Make sure your .pdb and/or .cif files are uploaded before running this block.\n",
        "\n",
        "%%script bash\n",
        "\n",
        "mkdir -p results models\n",
        "\n",
        "tar -xzf *tar.gz > /dev/null 2>&1\n",
        "\n",
        "# Delete any PDB or mmCIF files downloaded from the US-align repository.\n",
        "# US-align comes with 2 example PDB files and we don't want to add these to our tree.\n",
        "rm USalign_git/*.pdb\n",
        "\n",
        "# Move all pdb and cif files to the models directory to make file handling with colabalign.py a lot easier\n",
        "find . -type f \\( -name \"*.pdb\" -o -name \"*.cif\" \\) | tee /content/results/models.txt | xargs -I {} mv {} models/ > /dev/null 2>&1\n",
        "\n",
        "# Count the number of models in models/ to show how many alignments are needed.\n",
        "# Mainly to estimate how long the job will take.\n",
        "model_count=$(find models/ -type f | wc -l)\n",
        "calc_count=$((($model_count * ($model_count - 1)) / 2))\n",
        "echo \"There are ${model_count} files in the models directory. This means ${calc_count} total calculations.\"\n",
        "\n",
        "# Run colabalign.py with all available cores for the current runtype.\n",
        "python3 colabalign.py -i /content/models -o /content/results -c $(nproc --all) -t ${threshold} -p /content/USalign\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "e1XfdKzakiS7"
      },
      "outputs": [],
      "source": [
        "# @title Zip and download results\n",
        "\n",
        "import os\n",
        "import datetime\n",
        "from google.colab import files\n",
        "\n",
        "# Name the zipped results file using the current date and time\n",
        "# to not accidentally overwrite older results files when downloading\n",
        "current_dt = datetime.datetime.now()\n",
        "zip_filename = f'colabalign_results_{current_dt.strftime(\"%Y%m%d-%H%M\")}.zip'\n",
        "\n",
        "# Using the built-in zip function rather than a python module for improved efficiency\n",
        "os.system(f'zip -r {zip_filename} results')\n",
        "files.download(zip_filename)\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "TPU",
    "colab": {
      "gpuType": "V28",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "colabalign",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.12.7"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
