{"cells":[{"cell_type":"markdown","metadata":{"id":"Fm_wq8sBkiS3"},"source":["### TMmatrix&nbsp;&nbsp;&nbsp;&nbsp;[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/crfield18/TMmatrix/blob/main/tmmatrix.ipynb)\n","\n","Create a phylogenetic tree that compares the secondary structure of proteins, rather than nucleotide or amino acid sequence. Scoring uses the [_TM-align_](https://zhanggroup.org/TM-align/) algorithm by [_Zhang and Skolnick (2005)_](https://doi.org/10.1093%2Fnar%2Fgki524). Based of [_mTM-align_](http://yanglab.nankai.edu.cn/mTM-align/) by [_Dong et al. (2018)_](https://doi.org/10.1093/nar/gky430). Made possible with [Biopython](https://biopython.org).\n","\n","A score of **<0.17** indicates similarity indistinguishable from a random pair of structures, where as as score **â‰¥0.50** indicates a pair with broadly the same fold ([_Xu et al., 2010_](https://doi.org/10.1093/bioinformatics/btq066))\n","\n","\n","##### Usage\n","\n","* Upload .pdb or .cif format files directly to the Colab instance by clicking the folder icon on the left, then dragging and dropping your structures\n","\n","* For the best results, upload PDB/mmCIF files that contain a single subunit and with only a single position for each atom (crystallography/cryo-EM structures from the PDB can sometimes have multiple positions for some residues where a single position could not be resolved).\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"e8POztA5W7Qa"},"outputs":[],"source":["# @title Download and Compile the latest version of TM-align, Biopython and pandas\n","\n","%%script bash\n","\n","# Download Biopython and pandas\n","pip install biopython matplotlib\n","# pip install pandas # pandas is already installed on Colab. Uncomment if this changes\n","\n","# Download an up-to-date copy of TMalign/USalign\n","git clone https://github.com/pylelab/USalign.git\n","\n","# Compile TMalign\n","g++ -static -O3 -ffast-math -lm -o TMalign USalign/TMalign.cpp"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"FGAXnC9jXR6P"},"outputs":[],"source":["# @title Run pairwise TM-align calculations\n","\n","%%script bash\n","\n","# Make results and models directories\n","mkdir -p results models\n","\n","# Delete any PDB or mmCIF files downloaded from the USalign (TMalign) repository\n","rm USalign/*.pdb\n","\n","# Move all PDB and mmCIF files to /contents/models/ (silently)\n","find . -type f \\( -name \"*.pdb\" -o -name \"*.cif\" \\) -exec mv {} models/ \\; > /dev/null 2>&1\n","\n","# How many TM-align calculations are needed?\n","model_count=$(find models/ -type f | wc -l)\n","calc_count=$((($model_count * ($model_count - 1)) / 2))\n","\n","echo \"There are ${model_count} files in the models directory. This means ${calc_count} total calculations.\"\n","\n","# Make a list of all PDB and mmCIF files in the models directory and write to model_list.txt in the results directory\n","find models/ -maxdepth 1 \\( -name \"*.pdb\" -o -name \"*.cif\" \\) -printf \"%f\\n\" > results/model_list.txt\n","\n","# Run pairwise TM-align calculations\n","./TMalign -dir /content/models/ /content/results/model_list.txt -outfmt 2 > results/tmalign_out.tsv\n","# ./TMalign -h # Check that TMalign compiled correctly by loading the help message\n","# ./TMalign -dir /content/models/ /content/results/model_list.txt -outfmt 2 | tee results/tmalign_out.tsv # Print TM-align output to stdout (for debugging)"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"R7_g8hoBxOdO"},"outputs":[],"source":["# @title Make phylogenetic tree from TM-align results\n","\n","%matplotlib inline\n","\n","from math import isnan\n","from pathlib import Path\n","import pandas as pd\n","import numpy as np\n","\n","\n","from Bio import Phylo\n","from Bio.Phylo.TreeConstruction import DistanceTreeConstructor, DistanceMatrix\n","\n","class TMmatrix():\n","  def __init__(self) -> None:\n","    # Set home, models and results paths\n","    self.home_path = Path.cwd()\n","    self.models_path = self.home_path.joinpath('models')\n","    self.results_path = self.home_path.joinpath('results')\n","\n","    # Initialise self.tmmatrix as nothing\n","    self.tmmatrix = None\n","\n","    # Get model list from results/model_list.txt\n","    with open(self.results_path.joinpath('model_list.txt'), 'r') as model_list_file:\n","      self.model_list = [line.strip('\\n') for line in model_list_file]\n","\n","  def tmalign_output_to_matrix(self):\n","    # Load tmalign_out.tsv into self.tmalign_output\n","    tmalign_output = pd.read_csv(self.results_path.joinpath('tmalign_out.tsv'), sep='\\t')\n","\n","    # make new df with model list as column and rows\n","    self.tmmatrix = pd.DataFrame(index=self.model_list, columns=self.model_list)\n","\n","    # populate from output df with the highest TM-align score for each pair\n","    for index, row in tmalign_output.iterrows():\n","      if row['#PDBchain1'].startswith('#'):\n","        continue\n","      else:\n","        max_tm = max(row['TM1'], row['TM2'])\n","\n","        self.tmmatrix.at[row['#PDBchain1'], row['PDBchain2']] = max_tm\n","        self.tmmatrix.at[row['PDBchain2'], row['#PDBchain1']] = max_tm\n","\n","    # Fill the diagonal with scores of 1.0 (equivalent to aligning a model to itself)\n","    for obj in self.model_list:\n","      self.tmmatrix.at[obj,obj] = 1.0\n","\n","    # Write self.tmmatrix to tm-align_score_matrix.csv\n","    self.tmmatrix.to_csv(self.results_path.joinpath('tm-align_score_matrix.csv'), index=True)\n","\n","    return self.tmmatrix\n","\n","  def make_tree_from_matrix(self):\n","    # Invert TM-align scores\n","    # More similar pairs of models (i.e., higher scores) have shorter distances to each other\n","    if not self.tmmatrix.empty:\n","      distances_df = 1 - self.tmmatrix\n","\n","    # Convert the scores matrix to a lower triangle matrix\n","    lower_tri_df = distances_df.where(np.tril(np.ones(distances_df.shape)).astype(bool))\n","    lower_tri_lists = [[value for value in row if not isnan(value)] for row in lower_tri_df.values.tolist()]\n","\n","    # Generate phylogenetic tree using the UPGMA clustering method\n","    tm_matrix = DistanceMatrix(names=self.model_list, matrix=lower_tri_lists)\n","    constructor = DistanceTreeConstructor()\n","    tree = constructor.upgma(tm_matrix)\n","\n","    print('\\n')\n","\n","    # # Draw the tree for quick validation\n","    # # Only label terminal nodes\n","    # def no_internal_labels(node):\n","    #   if node.is_terminal():\n","    #     return node.name\n","    #   else:\n","    #     return None\n","\n","    # Phylo.draw(tree, label_func=no_internal_labels)\n","\n","    # Draw tree in ASCII for quick validation\n","    Phylo.draw_ascii(tree)\n","\n","    # Write the tree to results/tmmatrix.tree\n","    Phylo.write(tree, self.results_path.joinpath('tmmatrix.tree'), 'newick')\n","\n","  # Make a heatmap of TM-align values (unused by default)\n","  def tmmatrix_heatmap(self):\n","    import matplotlib.pyplot as plt\n","\n","    # Convert DataFrame to numeric values\n","    tmmatrix_numeric = self.tmmatrix.apply(pd.to_numeric, errors='coerce')\n","\n","    # Create the heatmap\n","    plt.figure(figsize=(25, 20))  # Adjust the size as needed\n","    plt.imshow(tmmatrix_numeric, cmap='cividis', interpolation='nearest')\n","\n","    # Set tick labels for x and y axes\n","    plt.xticks(ticks=range(len(tmmatrix_numeric.columns)), labels=tmmatrix_numeric.columns, rotation=90)\n","    plt.yticks(ticks=range(len(tmmatrix_numeric.index)), labels=tmmatrix_numeric.index)\n","\n","    # Add a colorbar\n","    plt.colorbar(label='TM-align score')\n","\n","    # Save and show the heatmap\n","    plt.savefig(self.results_path.joinpath('tmmatrix_heatmap.png'), dpi=300, bbox_inches='tight')\n","    plt.show()\n","\n","if __name__ == '__main__':\n","  colab_instance = TMmatrix()\n","  colab_instance.tmalign_output_to_matrix()\n","  colab_instance.make_tree_from_matrix()\n","  # colab_instance.tmmatrix_heatmap() # Uncomment to create a score heatmap"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"e1XfdKzakiS7"},"outputs":[],"source":["# @title Zip and download results\n","\n","import datetime\n","from google.colab import files\n","import zipfile\n","\n","current_dt = datetime.datetime.now()\n","\n","results_path = Path.cwd().joinpath('results')\n","zip_path = results_path.joinpath(f'tmmatrix_results_{current_dt.strftime(\"%Y%m%d-%H%M\")}.zip')\n","\n","with zipfile.ZipFile(zip_path, 'w', zipfile.ZIP_DEFLATED) as zipf:\n","  for file in results_path.rglob('*'):\n","    # Add only files (not directories) to the ZIP archive\n","    if file.is_file() and file.name != zip_path.name:\n","      zipf.write(file, arcname=file.relative_to(results_path))\n","\n","files.download(zip_path)"]}],"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/crfield18/TMmatrix/blob/main/tmmatrix.ipynb","timestamp":1722006336643}]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.12.2"}},"nbformat":4,"nbformat_minor":0}
