{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TM-align Phylogenetic Tree Maker (working title)\n",
    "\n",
    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/crfield18/TMmatrix/blob/main/tmmatrix.ipynb)\n",
    "\n",
    "\n",
    "Create a phylogenetic tree that compares the secondary structure of proteins, rather than nucleotide or amino acid sequence. Scoring uses the [_TM-align_](https://zhanggroup.org/TM-align/) algorithm by [_Zhang and Skolnick (2005)_](https://doi.org/10.1093%2Fnar%2Fgki524).\n",
    "\n",
    "A score of **<0.17** indicates similarity indistinguishable from a random pair of structures, where as as score **≥0.50** indicates a pair with the same fold ([_Xu et al., 2010_](https://doi.org/10.1093/bioinformatics/btq066)).\n",
    "\n",
    "\n",
    "Made possible with the use of: [Open-Source PyMol](https://github.com/schrodinger/pymol-open-source), [psico](https://github.com/speleo3/pymol-psico) and [Biopython](https://biopython.org).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Usage\n",
    "\n",
    "* Currently only working on _Google Chrome_ (I've had issues with the file upload step on _Safari_)\n",
    "\n",
    "* Run each code block individually by clicking the play ▶ icon.\n",
    "\n",
    "* Running the _Install Dependencies_ block will cause the kernal to restart, and an error stating _\"Your session crashed for an unknown reason\"_. This can be ignored.\n",
    "\n",
    "* If you need to restart, click _Runtime_ > _Disconnect and delete runtime_, then run each code block from the top of the page.\n",
    "\n",
    "* **RECOMMENDED MAXIMUM # of models** is 125. This equates to 7750 TM-align calucations (eq. 1), which should take ~ 1 hour to complete. You are welcome to bypass this limit, just be aware that the number of TM-align calculations increases by _n - 1_ where _n_ is the number of PDB/mmCIF files.\n",
    "\n",
    "* For the best results, upload PDB/mmCIF files that contain a single subunit.\n",
    "\n",
    "> Equation 1. $$\\frac{n!}{2(n-2)!}$$\n",
    "\n",
    "* Some PDB/mmCIF files (typically those generate from crystallographic data) can contain multiple locations for the same atom/residue (column 17 in PDB files). This can affect the TM-align score normalisation. For the highest accuracy, please only upload PDB/mmCIF files with 1 location for each atom."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Upload PDB/mmCIF files\n",
    "\n",
    "# @markdown Alternatively, click the folder icon on the left, `right click` > `New Folder` and name it _models_ (all lower case), then drag and drop your PDB/mmCIF files into the _models folder_.\n",
    "\n",
    "# Upload PDB/mmCIF files\n",
    "from google.colab import files\n",
    "\n",
    "uploaded = files.upload()\n",
    "\n",
    "# Create subdirectory for PDB/mmCIF files and results\n",
    "from pathlib import Path\n",
    "\n",
    "cwd = Path.cwd()\n",
    "for sub_dir in (cwd.joinpath('models'), cwd.joinpath('results')):\n",
    "  sub_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# Move PDB/mmCIF files to models subdirectory\n",
    "import shutil\n",
    "\n",
    "models_set = set(cwd.glob('*.pdb')).union(set(cwd.glob('*.cif')))\n",
    "\n",
    "for file in models_set:\n",
    "  source = file\n",
    "  dest = cwd.joinpath(f'models/{file.name}')\n",
    "  shutil.move(source, dest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Install Dependencies\n",
    "\n",
    "# @markdown This step installs the neccesary modules for python and pymol for this colab notebook to work.\n",
    "\n",
    "# @markdown _N.B._ **NOTHING** is installed locally on your computer. This all occurs on Google's servers.\n",
    "\n",
    "!pip install -q condacolab\n",
    "\n",
    "import condacolab\n",
    "condacolab.install()\n",
    "\n",
    "!mamba install conda-forge::pymol-open-source schrodinger::pymol-psico pandas conda-forge::biopython bioconda::tmalign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Calculate TM-align Matrix\n",
    "\n",
    "# @markdown This is where each model is compared with each other model, using the _TM-align_ algorithm. The total number of calculations needed to compate each model can be calculated using [Equation 1](#Usage).\n",
    "\n",
    "# @markdown Results should be downloaded automatically. If they don't, click the folder icon on the left, then `right click` > `Download` on results.zip.\n",
    "\n",
    "import sys, io\n",
    "from itertools import combinations\n",
    "from math import isnan\n",
    "from pathlib import Path\n",
    "import tqdm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pymol import cmd\n",
    "import psico.fullinit # Needed to add TM-align to PyMOL\n",
    "\n",
    "from Bio import Phylo\n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor, DistanceMatrix\n",
    "\n",
    "class TMalignMatrix():\n",
    "  def __init__(self) -> None:\n",
    "    self.models = {}\n",
    "    self.home_path = Path.cwd()\n",
    "    self.models_path = self.home_path.joinpath('models')\n",
    "    self.results_path = self.home_path.joinpath('results')\n",
    "    self.tmmatrix = pd.DataFrame()\n",
    "\n",
    "  def get_models(self):\n",
    "    models_set = set(self.models_path.glob('*.pdb')).union(set(self.models_path.glob('*.cif')))\n",
    "    for model in models_set:\n",
    "      self.models[model.name] = model\n",
    "    return self.models\n",
    "\n",
    "  def load_models(self):\n",
    "    if self.models == {}:\n",
    "      self.get_models()\n",
    "    for model, path in self.models.items():\n",
    "      # Load each model into pymol. Objects are named the same as the model file without .pdb/.cif\n",
    "      cmd.load(path, model[:-4])\n",
    "\n",
    "    # Remove any waters\n",
    "    cmd.remove('resn hoh')\n",
    "\n",
    "  def get_sequence_length(self, object_name):\n",
    "    # Count the number of CA atoms in given object\n",
    "    sequence_length = cmd.count_atoms(f'{object_name} and name CA')\n",
    "    return sequence_length\n",
    "\n",
    "  def calculate_matrix(self):\n",
    "    # Load all PDB/mmCIF files into pymol\n",
    "    self.load_models()\n",
    "    model_list = [model[:-4] for model, path in self.models.items() if model.endswith('.pdb') or model.endswith('.cif')]\n",
    "    print(f'Number of models: {len(model_list)}\\n')\n",
    "    # Initialise an empty pandas dataframe using the list of loaded pdb files as column/row indeces\n",
    "    if self.tmmatrix.empty:\n",
    "      self.tmmatrix = pd.DataFrame(index=model_list, columns=model_list)\n",
    "\n",
    "    # Calculate TM-align scores for all unique combinations of models\n",
    "    model_combos = list(combinations(model_list, 2))\n",
    "\n",
    "    for pair in tqdm.tqdm(model_combos, unit='calculations'):\n",
    "    # for pair in model_combos:\n",
    "      if isnan(self.tmmatrix.loc[pair[0], pair[1]]) or isnan(self.tmmatrix.loc[pair[1], pair[0]]):\n",
    "        # Set the smaller object as the reference (target) object\n",
    "        # TM-align scores are normalised based on sequence length\n",
    "        object_1_length = self.get_sequence_length(pair[0])\n",
    "        object_2_length = self.get_sequence_length(pair[1])\n",
    "\n",
    "        if object_1_length >= object_2_length:\n",
    "            target = pair[1]\n",
    "            mobile = pair[0]\n",
    "        elif object_1_length < object_2_length:\n",
    "            target = pair[0]\n",
    "            mobile = pair[1]\n",
    "\n",
    "        old_stdout = sys.stdout\n",
    "        captured_output = io.StringIO()\n",
    "        sys.stdout = captured_output\n",
    "\n",
    "        tm = cmd.tmalign(mobile, target)\n",
    "        self.tmmatrix.loc[pair[0], pair[1]] = tm\n",
    "        self.tmmatrix.loc[pair[1], pair[0]] = tm\n",
    "\n",
    "        sys.stdout = old_stdout\n",
    "        with open(self.results_path.joinpath('tm-align.log'), 'a+') as log:\n",
    "          log.write(captured_output.getvalue())\n",
    "\n",
    "    # Fill the diagonal with TM-align scores of 1 (always the score for identical proteins)\n",
    "    # ∴ We do not need to run any calculations\n",
    "    for obj in model_list:\n",
    "      self.tmmatrix.loc[obj,obj] = 1.0\n",
    "\n",
    "    # Write the TM-align score matrix to a csv file\n",
    "    self.tmmatrix.to_csv(self.results_path.joinpath('tm-align_score_matrix.csv'), index=True)\n",
    "\n",
    "    return self.tmmatrix\n",
    "\n",
    "  def make_tree_from_matrix(self):\n",
    "    if self.tmmatrix.empty:\n",
    "      distances_df = pd.read_csv(self.results_path.joinpath('tm-align_score_matrix.csv'), index_col=0)\n",
    "    else:\n",
    "      distances_df = self.tmmatrix\n",
    "\n",
    "    # Invert TM-align scores\n",
    "    # More similar pairs of models (i.e., higher scores) have shorter distances to each other\n",
    "    distances_df = 1 - distances_df\n",
    "\n",
    "    # Convert the scores matrix to a lower triangle matrix\n",
    "    lower_tri_df = distances_df.where(np.tril(np.ones(distances_df.shape)).astype(bool))\n",
    "    lower_tri_lists = [[value for value in row if not isnan(value)] for row in lower_tri_df.values.tolist()]\n",
    "\n",
    "    # Generate phylogenetic tree\n",
    "    tm_matrix = DistanceMatrix(names=distances_df.index.values.tolist(),matrix=lower_tri_lists)\n",
    "    constructor = DistanceTreeConstructor()\n",
    "    tree = constructor.upgma(tm_matrix)\n",
    "\n",
    "    # Draw the tree for quick validation\n",
    "    # Only label terminal nodes\n",
    "    def no_internal_labels(node):\n",
    "      if node.is_terminal():\n",
    "        return node.name\n",
    "      else:\n",
    "        return None\n",
    "    print('\\n')\n",
    "    Phylo.draw(tree, label_func=no_internal_labels)\n",
    "    # Phylo.draw_ascii(tree)\n",
    "\n",
    "    # Write the tree to results/results.tree\n",
    "    Phylo.write(tree, self.results_path.joinpath('results.tree'), 'newick')\n",
    "\n",
    "    # Zip and Dowload results directory\n",
    "    from google.colab import files\n",
    "    import zipfile\n",
    "\n",
    "    zip_path = self.home_path.joinpath('results.zip')\n",
    "\n",
    "    with zipfile.ZipFile(zip_path, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
    "      for file in self.results_path.rglob('*'):\n",
    "        if file.is_file() and file.name != 'results.zip':  # Add only files (not directories) to the ZIP archive\n",
    "          zipf.write(file, arcname=file.relative_to(self.results_path))\n",
    "\n",
    "    files.download(zip_path)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "  colab_instance = TMalignMatrix()\n",
    "  colab_instance.calculate_matrix()\n",
    "  colab_instance.make_tree_from_matrix()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
